### PR TITLE
Ensure Cache-Control: no-store on page-data files

### PR DESCRIPTION
**Pull request by lorisdev**
## Description

This PR sets the Cache-Control request header of the cache-dir loader to no-control ensuring that the cache-dir files are not cached in the browser preventing hard to find bugs.

### Documentation

The [documentation](https://www.gatsbyjs.com/docs/caching/) already states that those files should be hosted with Cache-Control: public, max-age=0, must-revalidate.
This PR adds another level of resilience in case you forgot to set the correct headers on your server / CDN.

## Related Issues

I didn't find a specific issue, but I got bitten myself by a stale cache. I can open an issue if needed.

### CODE REVIEW

**Comment by wardpeet** [comment]
@lorisdev can you share an example because I've seen that this isn't enough for many CDN cause it's cached on the edge.

**Comment by lorisdev** [comment]
@wardpeet from what I understand the "no-store" request directive is intended as an instruction to the browser to not cache or get from cache the response to this request.
It won't change the response headers sent back from the CDN but the thanks to this directive the files should be refetched from the CDN instead of the cache.

But you are right, I am not 100% sure it's failsafe, I try to setup an example to prove it.

File: packages/gatsby/cache-dir/loader.js [hunk 1]
```diff
@@ -38,6 +38,10 @@ function doFetch(url, method = `GET`) {
         resolve(req)
       }
     }
+    // Ensure responses are not cached. Usefull if page-data.json files are accidentally cached for example
+    if (noCache) {
+      req.setRequestHeader(`Cache-Control`, `no-store`)
```
Comment (wardpeet) [inline packages/gatsby/cache-dir/loader.js:43]: Let's move to no-cache so we still use disk-cache when etag and other validation methods are valid.

```suggestion
      req.setRequestHeader(`Cache-Control`, `no-cache`)
```

**Review by wardpeet — COMMENTED** [review]
I think this is a good addition, I'll give it a test to make sure cache-control is correct.

**Review by wardpeet — COMMENTED** [review]
I think this is a good addition, I'll give it a test to make sure cache-control is correct.

**Comment by lorisdev** [comment]
Yes `no-cache` should work as well.

In the meantime I have been playing with the cache header and the browser (tried on Chrome) doesn't seem to respect the directive with XMLHttpRequest but it works nicely with fetch API.

Basic example :
```
<!DOCTYPE html>
<html>
  <head>
    <title>Cache no-store repro</title>
    <meta charset="UTF-8" />
  </head>

  <body>
      <button id="XMLHttpRequest">XMLHttpRequest no-store</button>
      <button id="fetch">fetch no-store</button>

    <script type='module'>
        const reactCdnUrl = "https://unpkg.com/react@16.13.1/umd/react.production.min.js";

        const loadReactBundleOnCDN = async () => {
          try {
              await fetch(reactCdnUrl);
              console.log("React Bundle is loaded");
           } catch (e) {
              console.error(e);
          }
        };

        loadReactBundleOnCDN();

        const xmlHttpRequestButton = document.querySelector("#XMLHttpRequest");
        const fetchButton = document.querySelector("#fetch");

        xmlHttpRequestButton.addEventListener("click", () => {
           const client = new XMLHttpRequest();
           client.open("GET", reactCdnUrl, true);
           client.setRequestHeader("Cache-Control", "no-store");
           client.send(null);
        });

        fetchButton.addEventListener("click", () => {
           fetch(reactCdnUrl, { cache: "no-store" });
        });
    </script>
  </body>
</html>
```

You can load this html file in your browser and inspect the devtools.
At the first loading the file will be stored in cache.

Then you click on XMLHttpRequest button and the request will be fetched from cache

If you try with fetch button the request won't be fetched from cache.

cache-no-store

**Comment by lorisdev** [comment]
@wardpeet You have any idea why it makes the e2e_tests_path-prefix fail ?

**Comment by wardpeet** [comment]
sorry for the delay, i'll have a look

**Comment by KyleAMathews** [comment]
Doesn't this mean that the browser will just keep fetching over and over the same file? We really wouldn't want that. `must-revalidate` is much cheaper.

**Comment by KyleAMathews** [comment]
Another thought is we can test the cache-control headers coming back and if they're wrong, we then set out own like this or use cache busting query strings to ensure we're getting a new version.

**Review by wardpeet — COMMENTED** [review]
Hey, i've just tested the PR and we never hit the network through these functions because everything is preloaded.

So I'm going to close this one as it doesn't bring much value, only complexity. Thanks for all your time working on this. I'm sorry we couldn't merge this PR.

**Pull request closed without merging**